src/ForeignKeyDelegate.cpp

```diff
@@ -0,0 +1,156 @@
+        });
+
+        connect(tablesComboBox, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+                [=](int)
+        {
+            idsComboBox->setCurrentIndex(-1);
+        });
+
+        connect(idsComboBox, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+                [=](int index)
+        {
+            bool enableClausesEditor = (index!= -1);
+            m_clauseEdit->setEnabled(enableClausesEditor);
+        });
+    }
+
+    QString getResult() const
+    {
+        if (tablesComboBox->currentText().isEmpty() || idsComboBox->currentText().isEmpty())
+            return QString();
+        else
+            return QString("`%1`(`%2`) %3")
+                .arg(tablesComboBox->currentText())
+                .arg(idsComboBox->currentText())
+                .arg(m_clauseEdit->text())
+                .trimmed();
+    }
+
+    void setClause(const QString& clause)
+    {
+        m_clauseEdit->setText(clause);
+    }
+
+    QComboBox* tablesComboBox;
+    QComboBox* idsComboBox;
+
+private:
+    QPushButton* m_btnReset;
+    QLineEdit* m_clauseEdit; // for ON CASCADE and such
+};
+
+ForeignKeyDelegate::ForeignKeyDelegate(const DBBrowserDB& db, sqlb::Table& table, QObject* parent)
+    : QStyledItemDelegate(parent)
+    , m_db(db)
+    , m_table(table)
+{
+
+}
+
+QWidget* ForeignKeyDelegate::createEditor(QWidget* parent, const QStyleOptionViewItem& option, const QModelIndex& index) const
+{
+    Q_UNUSED(option)
+    Q_UNUSED(index)
+
+    ForeignKeyEditor* editor = new ForeignKeyEditor(parent);
+    editor->setAutoFillBackground(true);
+
+    // get tables list
+    QStringList list;
+    auto objects = m_db.getBrowsableObjects();
+    for (auto obj : objects) {
+        if ("table" == obj.gettype()) {
+            QString tableName = obj.table.name();
+            if (tableName == m_table.name())
+                continue;
+
+            m_tablesIds.insert(tableName, obj.table.fieldNames());
+        }
+    }
+
+    editor->tablesComboBox->addItems(m_tablesIds.keys());
+    editor->tablesComboBox->setCurrentIndex(-1);
+
+    connect(editor->tablesComboBox, static_cast<void(QComboBox::*)(const QString &)>(&QComboBox::currentIndexChanged),
+            [=](const QString& tableName)
+    {
+        QComboBox* box = editor->idsComboBox;
+        box->clear();
+        box->addItems(m_tablesIds[tableName]);
+        box->setCurrentIndex(-1);
+    });
+
+    return editor;
+}
+
+void ForeignKeyDelegate::setEditorData(QWidget* editor, const QModelIndex& index) const
+{
+    ForeignKeyEditor* fkEditor = static_cast<ForeignKeyEditor*>(editor);
+
+    // Doing regex in this case is a bit weird, yeah. Reason for this
+    // is that we can grab table(), columns() and constraint() only
+    // from foreign keys that had actually passed thru lexer. For a
+    // brand new foreign key those methods return empty fields.
+    QRegularExpression re("\\`(\\w*)\\`\\(\\`(\\w*)\\`\\)(.*)");
```

Comment:
Probably needs to handle quoting of ``` ` ``` quotes correctly, too :frowning: Might become ugly.